In [1]:
import pandas as pd

## Arabidopsis

In [2]:
Supp_4_df = pd.read_excel (r'Supplemental_file_4[1].xls')

In [3]:
Supp_4_df

The structural features of Arabidopsis flowering genes  \
0                                               GeneID       
1                                            AT1G01060       
2                                            AT1G01640       
3                                            AT1G02400       
4                                            AT1G04400       
..                                                 ...       
200                                          AT5G65050       
201                                          AT5G65060       
202                                          AT5G65070       
203                                          AT5G65080       
204                                          AT5G67100       

                Unnamed: 1               Unnamed: 2           Unnamed: 3  \
0    Number of transcripts  Average number of exons  Average exon length   
1                        5                        9                296.7   
2                        2                        3                299.2   
3                        1                        3                445.3   
4                        2                        5                469.8   
..                     ...                      ...                  ...   
200                      3                      6.3                142.7   
201                      2                        7                125.7   
202                      3                      7.7                121.5   
203                      2                      6.5                124.6   
204                      1                       30                151.5   

                Unnamed: 4  
0    Average intron length  
1                    983.4  
2                    636.8  
3                      720  
4                   1064.3  
..                     ...  
200                  710.3  
201                  775.6  
202                  693.7  
203                 1040.7  
204                  122.9  

[205 rows x 5 columns]

First tab from supplemental file 4 uploaded (Arabidopsis)

In [14]:
Sorted_transcript_data = Supp_4_df.groupby('Unnamed: 1')

In [19]:
Sorted_transcript_data.describe()

The structural features of Arabidopsis flowering genes  \
                                                                       count   
Unnamed: 1                                                                     
1                                                                    151       
2                                                                     35       
3                                                                     12       
4                                                                      5       
5                                                                      1       
Number of transcripts                                                  1       

                                             Unnamed: 2         \
                      unique        top freq      count unique   
Unnamed: 1                                                       
1                        151  AT1G02400    1        151     21   
2                         35  AT1G01640    1       35.0   19.0   
3                         12  AT1G24260    1       12.0   10.0   
4                          5  AT1G77080    1        5.0    5.0   
5                          1  AT1G01060    1          1      1   
Number of transcripts      1     GeneID    1          1      1   

                                                    Unnamed: 3         \
                                           top freq      count unique   
Unnamed: 1                                                              
1                                            1   37      151.0  148.0   
2                                          2.5  5.0       35.0   35.0   
3                                          7.7  2.0       12.0   12.0   
4                                          6.2  1.0        5.0    5.0   
5                                            9    1        1.0    1.0   
Number of transcripts  Average number of exons    1          1      1   

                                                Unnamed: 4         \
                                       top freq      count unique   
Unnamed: 1                                                          
1                                    119.2  2.0      151.0  114.0   
2                                    299.2  1.0       35.0   35.0   
3                                    138.0  1.0       12.0   12.0   
4                                    122.0  1.0        5.0    5.0   
5                                    296.7  1.0        1.0    1.0   
Number of transcripts  Average exon length    1          1      1   

                                                    
                                         top  freq  
Unnamed: 1                                          
1                                        0.0  37.0  
2                                      636.8   1.0  
3                                      491.3   1.0  
4                                      726.8   1.0  
5                                      983.4   1.0  
Number of transcripts  Average intron length     1

In [23]:
(5+(4*5)+(3*12)+(2*35)+151)/204

1.3823529411764706

Number of transcripts per gene. 1.38 calculated here as compared to 1.4 that the authors recieved. Range is 1-5 as seen above in the sorted data.

In [27]:
Sorted_exon_data = Supp_4_df.groupby('Unnamed: 2')

In [28]:
Sorted_exon_data.describe()

The structural features of Arabidopsis flowering genes  \
                                                                         count   
Unnamed: 2                                                                       
1                                                                       37       
1.5                                                                      2       
2                                                                       18       
2.5                                                                      5       
2.7                                                                      1       
3                                                                       24       
4                                                                       12       
4.3                                                                      1       
5                                                                        9       
5.5                                                                      2       
6                                                                       14       
6.2                                                                      1       
6.3                                                                      1       
6.5                                                                      2       
7                                                                       10       
7.5                                                                      1       
7.7                                                                      2       
8                                                                        8       
8.3                                                                      2       
8.7                                                                      1       
9                                                                        8       
9.3                                                                      1       
9.5                                                                      1       
10                                                                       5       
11                                                                       4       
12                                                                       2       
13                                                                       3       
13.5                                                                     1       
14                                                                       7       
14.5                                                                     2       
15                                                                       4       
17                                                                       2       
17.5                                                                     1       
18                                                                       1       
18.5                                                                     1       
20                                                                       2       
20.5                                                                     1       
24                                                                       1       
29                                                                       1       
30                                                                       1       
34.7                                                                     1       
48.3                                                                     1       
Average number of exons                                                  1       

                                               Unnamed: 1         \
                        unique        top freq      count unique   
Unnamed: 2                                                         
1                           37  AT1G12610    1 

In [35]:
(48.3+34.7+30+29+24+20+(20*2)+18.5+18+17.5+(17*2)+(15*4)+(14.5*2)+(14*7)+13.5+(13*2)+(12*2)+(11*4)+(10*5)+9.5+9.3+8.7+(9*8)+(8.3*2)+(8*8)+(7.7*2)+7.5+(7*10)+6.3+(6.5*2)+6.2+(6*14)+(5.5*2)+(5*9)+4.3+(4*12)+2.7+(3*24)+(2.5*5)+(2*18)+(1.5*2)+(1*37))/204


6.4338235294117645

6.433 Exons per gene. Range is 1-48.3

In [37]:
Sorted_Esize_data = Supp_4_df.groupby('Unnamed: 3')

In [38]:
Sorted_Esize_data.describe()

The structural features of Arabidopsis flowering genes  \
                                                                     count   
Unnamed: 3                                                                   
79                                                                   1       
80.3                                                                 1       
81                                                                   1       
84.6                                                                 1       
85                                                                   1       
...                                                                ...       
2128                                                                 1       
2301                                                                 1       
2712                                                                 1       
4165                                                                 1       
Average exon length                                                  1       

                                           Unnamed: 1         \
                    unique        top freq      count unique   
Unnamed: 3                                                     
79                       1  AT4G30972    1          1      1   
80.3                     1  AT1G18450    1          1      1   
81                       1  AT2G25095    1          1      1   
84.6                     1  AT2G06005    1          1      1   
85                       1  AT5G10945    1          1      1   
...                    ...        ...  ...        ...    ...   
2128                     1  AT1G14920    1          1      1   
2301                     1  AT2G01570    1          1      1   
2712                     1  AT1G62830    1          1      1   
4165                     1  AT4G39400    1          1      1   
Average exon length      1     GeneID    1          1      1   

                                                Unnamed: 2         \
                                       top freq      count unique   
Unnamed: 3                                                          
79                                       1    1          1      1   
80.3                                     1    1          1      1   
81                                       1    1          1      1   
84.6                                     2    1          1      1   
85                                       1    1          1      1   
...                                    ...  ...        ...    ...   
2128                                     1    1          1      1   
2301                                     1    1          1      1   
2712                                     1    1          1      1   
4165                                     1    1          1      1   
Average exon length  Number of transcripts    1          1      1   

                                                  Unnamed: 4         \
                                         top freq      count unique   
Unnamed: 3                                                            
79                                         1    1          1      1   
80.3                                      20    1          1      1   
81                                         1    1          1      1   
84.6                                      14    1        1.0    1.0   
85                                         1    1          1      1   
...                                      ...  ...        ...    ...   
2128                                       1    1          1      1   
2301                                       1    1          1      1   
2712                                       1    1          1      1   
4165                                       1    1          1      1   
Average exon length  Average number of exons    1          1      1   

                                                 
                  

In [51]:
Supp_4_df.drop(0)

The structural features of Arabidopsis flowering genes Unnamed: 1  \
1                                            AT1G01060              5   
2                                            AT1G01640              2   
3                                            AT1G02400              1   
4                                            AT1G04400              2   
5                                            AT1G09570              2   
..                                                 ...            ...   
200                                          AT5G65050              3   
201                                          AT5G65060              2   
202                                          AT5G65070              3   
203                                          AT5G65080              2   
204                                          AT5G67100              1   

    Unnamed: 2 Unnamed: 3 Unnamed: 4  
1            9      296.7      983.4  
2            3      299.2      636.8  
3            3      445.3        720  
4            5      469.8     1064.3  
5          5.5      686.9     1987.7  
..         ...        ...        ...  
200        6.3      142.7      710.3  
201          7      125.7      775.6  
202        7.7      121.5      693.7  
203        6.5      124.6     1040.7  
204         30      151.5      122.9  

[204 rows x 5 columns]

In [58]:
Supp_4_copy

The structural features of Arabidopsis flowering genes  \
0                                               GeneID       
1                                            AT1G01060       
2                                            AT1G01640       
3                                            AT1G02400       
4                                            AT1G04400       
..                                                 ...       
200                                          AT5G65050       
201                                          AT5G65060       
202                                          AT5G65070       
203                                          AT5G65080       
204                                          AT5G67100       

                Unnamed: 1               Unnamed: 2           Unnamed: 3  \
0    Number of transcripts  Average number of exons  Average exon length   
1                        5                        9                296.7   
2                        2                        3                299.2   
3                        1                        3                445.3   
4                        2                        5                469.8   
..                     ...                      ...                  ...   
200                      3                      6.3                142.7   
201                      2                        7                125.7   
202                      3                      7.7                121.5   
203                      2                      6.5                124.6   
204                      1                       30                151.5   

                Unnamed: 4  
0    Average intron length  
1                    983.4  
2                    636.8  
3                      720  
4                   1064.3  
..                     ...  
200                  710.3  
201                  775.6  
202                  693.7  
203                 1040.7  
204                  122.9  

[205 rows x 5 columns]

In [55]:
Supp_4_copy = Supp_4_df

In [60]:
Supp_4_copy.drop('The structural features of Arabidopsis flowering genes', axis=1, inplace=True)

In [63]:
Supp_4_copy_1= Supp_4_copy.drop(labels=0, axis=0)

In [64]:
Supp_4_copy_1

Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4
1            5          9      296.7      983.4
2            2          3      299.2      636.8
3            1          3      445.3        720
4            2          5      469.8     1064.3
5            2        5.5      686.9     1987.7
..         ...        ...        ...        ...
200          3        6.3      142.7      710.3
201          2          7      125.7      775.6
202          3        7.7      121.5      693.7
203          2        6.5      124.6     1040.7
204          1         30      151.5      122.9

[204 rows x 4 columns]

In [65]:
Supp_4_df_num = Supp_4_copy_1.apply(pd.to_numeric)

In [66]:
Supp_4_df_num

Unnamed: 1  Unnamed: 2  Unnamed: 3  Unnamed: 4
1             5         9.0       296.7       983.4
2             2         3.0       299.2       636.8
3             1         3.0       445.3       720.0
4             2         5.0       469.8      1064.3
5             2         5.5       686.9      1987.7
..          ...         ...         ...         ...
200           3         6.3       142.7       710.3
201           2         7.0       125.7       775.6
202           3         7.7       121.5       693.7
203           2         6.5       124.6      1040.7
204           1        30.0       151.5       122.9

[204 rows x 4 columns]

In [67]:
Supp_4_copy_1['Unnamed: 3'].mean()

466.1058823529412

In [69]:
Supp_4_copy_1['Unnamed: 3'].min()

79

In [70]:
Supp_4_copy_1['Unnamed: 3'].max()

4165

Exon average size is 466. Range is 79-4165. This agrees with the authors data.

In [71]:
Supp_4_copy_1['Unnamed: 4'].mean()

468.4995098039214

In [73]:
Supp_4_copy_1['Unnamed: 4'].max()

2316

In [96]:
supp_4_desc = Supp_4_copy_1.sort_values('Unnamed: 4', ascending=False)

Intron average size was 468.5bp and range was 78-2316 bp. I had to sort data in order to determine the minimum number since the authors ignored the 0 values for introns.

In [104]:
supp_4_desc.iloc[165:170]

Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4
72           2          2      779.8       84.3
122          3          3      125.3       77.6
18           1          1       1364          0
101          1          1        425          0
191          1          1        124          0

In [116]:
for index, row in Supp_4_df_num.iterrows():
    exon = row['Unnamed: 3']
    numexon = row['Unnamed: 2']
    if(pd.isna(exon) is False and pd.isna(intron) is False): 
        Supp_4_df_num.loc[index,'exon_total'] = exon * numexon

In [120]:
for index, row in Supp_4_df_num.iterrows():
    exon_total = row['exon_total']
    intron = row['Unnamed: 4']
    if(pd.isna(exon) is False and pd.isna(intron) is False): 
        Supp_4_df_num.loc[index,'exon_intron'] = exon_total + intron

In [121]:
Supp_4_df_num

Unnamed: 1  Unnamed: 2  Unnamed: 3  Unnamed: 4   new_value  exon_intron  \
1             5         9.0       296.7       983.4   291774.78      3653.70   
2             2         3.0       299.2       636.8   190530.56      1534.40   
3             1         3.0       445.3       720.0   320616.00      2055.90   
4             2         5.0       469.8      1064.3   500008.14      3413.30   
5             2         5.5       686.9      1987.7  1365351.13      5765.65   
..          ...         ...         ...         ...         ...          ...   
200           3         6.3       142.7       710.3   101359.81      1609.31   
201           2         7.0       125.7       775.6    97492.92      1655.50   
202           3         7.7       121.5       693.7    84284.55      1629.25   
203           2         6.5       124.6      1040.7   129671.22      1850.60   
204           1        30.0       151.5       122.9    18619.35      4667.90   

     exon_total  
1       2670.30  
2        897.60  
3       1335.90  
4       2349.00  
5       3777.95  
..          ...  
200      899.01  
201      879.90  
202      935.55  
203      809.90  
204     4545.00  

[204 rows x 7 columns]

In [122]:
Supp_4_df_num['exon_intron'].mean()

2297.3894607843135

In [124]:
Supp_4_df_num['exon_intron'].min()

79.0

In [123]:
Supp_4_df_num['exon_intron'].max()

11046.510000000002

After making two new column to calculated gene length, different numbers were calculated from what the authors provided. Exon_total is the length of exons * the number they appear per transcript to get their total length. Then, Exon_total was added to intron length to compute gene length. There was not a good description of how the authors calculated this value. 
Average: 2297bp
Range: 79-11046.5bp

In [125]:
Supp_4_df_num['exon_total'].mean()

1828.8899509803912

In [126]:
Supp_4_df_num['exon_total'].max()

10871.510000000002

In [127]:
Supp_4_df_num['exon_total'].min()

79.0

Protein length was calculated by taking the number of exons multiplied by the average number per gene. 
Average:1828aa
Range:79-10871.5aa
These values are different than what was calculated by the authors. Again, there was no straightforward directions on how this was done. 

## Wheat